### PR TITLE
Enable docker metrics

### PR DESCRIPTION
**Pull request by davemachado**

### CODE REVIEW

**Comment by manojm321** [comment]
having this in default.conf is not a good idea because some of the machines do not run docker.

we can move this to a separate file telegraf-docker.conf and create a separate rpm. Please look at telegraf-redis.conf for an example.

Also you would need to update arista-package.sh to create a new docker rpm.

File: Arista/arista-package.sh [hunk 1]
```diff
@@ -126,6 +126,11 @@ rm -f $TMP_CONFIG_DIR/etc/telegraf/telegraf.d/*
 cp $CONFIG_FILES_DIR/telegraf-redis.conf $TMP_CONFIG_DIR/etc/telegraf/telegraf.d/
 fpm -s dir -t rpm $CONFIG_FPM_ARGS --description "$DESCRIPTION" -n "telegraf-Redis" etc lib || cleanup_exit 1
 
+# Docker-Config
```
Comment (manojm321) [inline Arista/arista-package.sh:125]: please increment CONFIG_FILES_ITER up top

File: Arista/ConfigFiles/telegraf-docker.conf [hunk 2]
```diff
@@ -0,0 +1,48 @@
+###############################################################################
```
Comment (manojm321) [inline Arista/ConfigFiles/telegraf-docker.conf:1]: output plugin is required.

File: Arista/ConfigFiles/telegraf-default.conf [hunk 3]
```diff
@@ -94,7 +94,6 @@
   ## Uncomment the following line if you do not need disk serial numbers.
   skip_serial_number = true
 
-
```
Comment (manojm321) [inline Arista/ConfigFiles/telegraf-default.conf]: nit: white space changes

File: Arista/arista-package.sh (hunk 1 above)
Comment (davemachado) [reply Arista/arista-package.sh:125]: Done.

File: Arista/ConfigFiles/telegraf-docker.conf (hunk 2 above)
Comment (davemachado) [reply Arista/ConfigFiles/telegraf-docker.conf:1]: Done.

**Comment by kthommandra** [comment]
Include a testing note, output of the -test command to ensure changes work fine when deployed.

Manoj has changed the versioning of Telegraf .. so kindly update this change accordingly.
Rest of the stuff looks fine to me.

**Comment by kthommandra** [comment]
any update on this ?

**Comment by davemachado** [comment]
Things look to be good now that some changes have been made to our docker config. Telegraf can now access the docker socket:

```
Arista/ConfigFiles » sudo telegraf --config telegraf-docker.conf -input-filter docker -test
* Plugin: inputs.docker, Collection 1
> docker,engine_host=us172.sjc.aristanetworks.com,host=us172.sjc.aristanetworks.com n_containers_paused=0i,n_listener_events=0i,n_cpus=40i,n_containers=221i,n_containers_running=65i,n_containers_stopped=156i,n_images=381i,n_goroutines=287i,n_used_file_descriptors=408i 1535594844000000000
> docker,unit=bytes,engine_host=us172.sjc.aristanetworks.com,host=us172.sjc.aristanetworks.com memory_total=269841690624i 1535594844000000000
> docker_container_mem,container_name=kduda.vendor.0,netMode=standalone-host,a4cBuildVersion=experimental,changeNum=9645493,ctnrType=project,owner=kduda,container_image=sha256,platform=i386_18,project=kduda.vendor.0,host=us172.sjc.aristanetworks.com,service=workspace,volumeMount=/var/docker/volumes/kduda.vendor.0/_data,timestamp=2018-08-28\ 23:32,nickname=vendor.0,volume=kduda.vendor.0,engine_host=us172.sjc.aristanetworks.com,container_version=0fd30c48646edbad1bc6372d44d4f7b987fa289546a375dc75cabec7b91ef1bd,a4cVersion=experimental total_inactive_anon=13860864i,total_inactive_file=2771828736i,total_pgfault=1703544i,inactive_file=2771828736i,mapped_file=4096i,rss=25600000i,total_cache=2819899392i,inactive_anon=13860864i,pgfault=1703544i,pgpgin=1586734i,total_rss_huge=0i,usage=2930814976i,max_usage=3094552576i,active_anon=11743232i,cache=2819899392i,total_mapped_file=4096i,total_writeback=0i,usage_percent=0.04110394644486231,pgmajfault=26i,rss_huge=0i,total_active_file=48066560i,total_unevictable=0i,container_id="c1037ea1e43d9808f25be062faaee4696d88fe93499699013bc5417b0f1f6312",pgpgout=892032i,total_pgmajfault=26i,hierarchical_memory_limit=9223372036854771712i,total_pgpgout=892032i,unevictable=0i,writeback=0i,limit=269841690624i,active_file=48066560i,total_active_anon=11743232i,total_pgpgin=1586734i,total_rss=25600000i 1535594845000000000
> docker_container_cpu,timestamp=2018-08-28\ 23:32,nickname=vendor.0,container_image=sha256,project=kduda.vendor.0,changeNum=9645493,container_version=0fd30c48646edbad1bc6372d44d4f7b987fa289546a375dc75cabec7b91ef1bd,a4cVersion=experimental,service=workspace,owner=kduda,host=us172.sjc.aristanetworks.com,volume=kduda.vendor.0,container_name=kduda.vendor.0,platform=i386_18,a4cBuildVersion=experimental,ctnrType=project,engine_host=us172.sjc.aristanetworks.com,volumeMount=/var/docker/volumes/kduda.vendor.0/_data,netMode=standalone-host,cpu=cpu-total usage_in_kernelmode=25860000000i,throttling_throttled_periods=0i,container_id="c1037ea1e43d9808f25be062faaee4696d88fe93499699013bc5417b0f1f6312",usage_percent=0.0014401401401401403,usage_total=49800162590i,usage_in_usermode=21420000000i,usage_system=22105363090000000i,throttling_periods=0i,throttling_throttled_time=0i 1535594845000000000
> docker_container_cpu,nickname=vendor.0,container_version=0fd30c48646edbad1bc6372d44d4f7b987fa289546a375dc75cabec7b91ef1bd,service=workspace,platform=i386_18,project=kduda.vendor.0,host=us172.sjc.aristanetworks.com,timestamp=2018-08-28\ 23:32,owner=kduda,a4cBuildVersion=experimental,changeNum=9645493,a4cVersion=experimental,container_image=sha256,volume=kduda.vendor.0,container_name=kduda.vendor.0,cpu=cpu0,ctnrType=project,engine_host=us172.sjc.aristanetworks.com,volumeMount=/var/docker/volumes/kduda.vendor.0/_data,netMode=standalone-host container_id="c1037ea1e43d9808f25be062faaee4696d88fe93499699013bc5417b0f1f6312",usage_total=1160521478i 1535594845000000000
> docker_container_cpu,volume=kduda.vendor.0,platform=i386_18,ctnrType=project,container_version=0fd30c48646edbad1bc6372d44d4f7b987fa289546a375dc75cabec7b91ef1bd,a4cVersion=experimental,volumeMount=/var/docker/volumes/kduda.vendor.0/_data,changeNum=9645493,host=us172.sjc.aristanetworks.com,engine_host=us172.sjc.aristanetworks.com,project=kduda.vendor.0,cpu=cpu1,service=workspace,netMode=standalone-host,owner=kduda,a4cBuildVersion=experimental,container_image=sha256,timestamp=2018-08-28\ 23:32,nickname=vendor.0,container_name=kduda.vendor.0 usage_total=1028661979i,container_id="c1037ea1e43d9808f25be062faaee4696d88fe93499699013bc5417b0f1f6312" 1535594845000000000
> docker_container_cpu,netMode=standalone-host,a4cBuildVersion=experimental,a4cVersion=experimental,container_version=0fd30c48646edbad1bc6372d44d4f7b987fa289546a375dc75cabec7b91ef1bd,engine_host=us172.sjc.aristanetworks.com,volume=kduda.vendor.0,container_name=kduda.vendor.0,owner=kduda,container_image=sha256,platform=i386_18,host=us172.sjc.aristanetworks.com,volumeMount=/var/docker/volumes/kduda.vendor.0/_data,timestamp=2018-08-28\ 23:32,nickname=vendor.0,changeNum=9645493,ctnrType=project,project=kduda.vendor.0,cpu=cpu2,service=workspace usage_total=529854784i,container_id="c1037ea1e43d9808f25be062faaee4696d88fe93499699013bc5417b0f1f6312" 1535594845000000000
> docker_container_cpu,nickname=vendor.0,engine_host=us172.sjc.aristanetworks.com,container_version=0fd30c48646edbad1bc6372d44d4f7b987fa289546a375dc75cabec7b91ef1bd,a4cVersion=experimental,netMode=standalone-host,project=kduda.vendor.0,a4cBuildVersion=experimental,ctnrType=project,cpu=cpu3,host=us172.sjc.aristanetworks.com,timestamp=2018-08-28\ 23:32,volume=kduda.vendor.0,changeNum=9645493,owner=kduda,container_image=sha256,volumeMount=/var/docker/volumes/kduda.vendor.0/_data,container_name=kduda.vendor.0,platform=i386_18,service=workspace usage_total=949743959i,container_id="c1037ea1e43d9808f25be062faaee4696d88fe93499699013bc5417b0f1f6312" 1535594845000000000
> docker_container_cpu,service=workspace,timestamp=2018-08-28\ 23:32,container_name=kduda.vendor.0,a4cBuildVersion=experimental,owner=kduda,cpu=cpu4,a4cVersion=experimental,volumeMount=/var/docker/volumes/kduda.vendor.0/_data,volume=kduda.vendor.0,host=us172.sjc.aristanetworks.com,container_version=0fd30c48646edbad1bc6372d44d4f7b987fa289546a375dc75cabec7b91ef1bd,project=kduda.vendor.0,ctnrType=project,container_image=sha256,platform=i386_18,engine_host=us172.sjc.aristanetworks.com,netMode=standalone-host,changeNum=9645493,nickname=vendor.0 usage_total=939780948i,container_id="c1037ea1e43d9808f25be062faaee4696d88fe93499699013bc5417b0f1f6312" 1535594845000000000
...
```

File: Arista/ConfigFiles/telegraf-docker.conf [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  ## When empty only containers in the "running" state will be captured.
+  # container_state_include = []
+  # container_state_exclude = []
+
+  ## Timeout for docker list, info, and stats commands
+  timeout = "5s"
+
+  ## Whether to report for each container per-device blkio (8:0, 8:1...) and
+  ## network (eth0, eth1, ...) stats or not
+  perdevice = true
```
Comment (kthommandra) [inline Arista/ConfigFiles/telegraf-docker.conf]: is this necessary? aren't there some tests that could run in containers and create lots of network interfaces? is it necessary to track per interface and per device stats? Isn't the total good enough?

File: Arista/ConfigFiles/telegraf-docker.conf [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  ## Whether to report for each container total blkio and network stats or not
+  total = false
+
+  ## docker labels to include and exclude as tags.  Globs accepted.
+  ## Note that an empty array for both will include all labels as tags
+  docker_label_include = []
+  docker_label_exclude = []
+
+  ## Which environment variables should we use as a tag
+  tag_env = ["JAVA_HOME", "HEAP_SIZE"]
```
Comment (kthommandra) [inline Arista/ConfigFiles/telegraf-docker.conf]: Just curious why do we need JAVA_HOME as tag? Did you intend to use something else here?

**Comment by kthommandra** [comment]
Question about the output from -test, the measurements like docker_container_mem and docker_container_cpu seem to have lot of high cardinality tags eg: change_num and container_version etc. Have you discussed the impact of such points being written to Influx with the team already?

File: Arista/ConfigFiles/telegraf-docker.conf (hunk 4 above)
Comment (davemachado) [reply Arista/ConfigFiles/telegraf-docker.conf]: Yes, good point. Updated.

File: Arista/ConfigFiles/telegraf-docker.conf (hunk 5 above)
Comment (davemachado) [reply Arista/ConfigFiles/telegraf-docker.conf]: I had removed these and updated the label filtering above but did not `git add` and `git push` the file... :-/ 

"Updated" to ignore k8s management labels and only include a4c and general workspace labels (containerType and platform, both of which have a cardinality < 5)

**Comment by davemachado** [comment]
> Have you discussed the impact of such points being written to Influx with the team already?

I have not. Does a high cardinality grind influx to a halt? I will start a conversation with the team about this.

**Comment by davemachado** [comment]
To start out, I'd like to use `namepass` to only record points for the `docker`  measurement. That single measurement contains most of the info we care about at this point in our usage of docker, and that avoids the high cardinality of the other measurements altogether. 

New output:

```
us172 :: Arista/ConfigFiles » sudo telegraf --config telegraf-docker.conf -test
* Plugin: inputs.docker, Collection 1
> docker,engine_host=us172.sjc.aristanetworks.com,host=us172.sjc.aristanetworks.com n_containers_paused=0i,n_goroutines=300i,n_listener_events=0i,n_used_file_descriptors=423i,n_containers=222i,n_containers_stopped=155i,n_images=382i,n_cpus=40i,n_containers_running=67i 1535639795000000000
> docker,unit=bytes,engine_host=us172.sjc.aristanetworks.com,host=us172.sjc.aristanetworks.com memory_total=269841690624i 1535639795000000000
```

**Comment by kthommandra** [comment]
LGTM, will let manoj take a look and do the merge

**Review by manojm321 — APPROVED** [review]
LGTM!

**Pull request merged**